contracts/src/v0.8/ccip/pools/SiloedLockReleaseTokenPool.sol

```diff
@@ -0,0 +1,181 @@
+/// @dev One token per LockReleaseTokenPool.
+contract SiloedLockReleaseTokenPool is TokenPool, ITypeAndVersion {
+  using SafeERC20 for IERC20;
+  using EnumerableSet for EnumerableSet.UintSet;
+
+  error InsufficientLiquidity();
+  error LiquidityNotAccepted();
+
+  event LiquidityTransferred(uint64 remoteChainSelector, address indexed from, uint256 amount);
+  event LiquidityAdded(uint64 remoteChainSelector, address indexed provider, uint256 indexed amount);
+  event LiquidityRemoved(uint64 remoteChainSelector, address indexed provider, uint256 indexed amount);
+  event ChainSiloeDesignationUpdated(uint64 remoteChainSelector, bool isSiloed);
+
+  event RebalancerSet(uint64 indexed remoteChainSelector, address oldRebalancer, address newRebalancer);
+
+  string public constant override typeAndVersion = "SiloedLockReleaseTokenPool 1.5.10-dev";
+
+  mapping(uint64 chainSelector => uint256 lockedBalance) internal s_lockedTokensByChainSelector;
+  mapping(uint64 remoteChainSelector => address rebalancer) internal s_rebalancerByChain;
+  mapping(uint64 remoteChainSelector => bool shouldSiloFunds) internal s_siloedChainSelectors;
+
+  EnumerableSet.UintSet internal s_siloedChains;
+
+  constructor(
+    IERC20 token,
+    uint8 localTokenDecimals,
+    address[] memory allowlist,
+    address rmnProxy,
+    address router
+  ) TokenPool(token, localTokenDecimals, allowlist, rmnProxy, router) {}
+
+  /// @notice Locks the token in the pool
+  /// @dev The _validateLockOrBurn check is an essential security check
+  function lockOrBurn(
+    Pool.LockOrBurnInV1 calldata lockOrBurnIn
+  ) external virtual override returns (Pool.LockOrBurnOutV1 memory) {
+    _validateLockOrBurn(lockOrBurnIn);
+
+    // If funds need to be siloed, update internal accounting;
+    if (s_siloedChainSelectors[lockOrBurnIn.remoteChainSelector]) {
+      s_lockedTokensByChainSelector[lockOrBurnIn.remoteChainSelector] += lockOrBurnIn.amount;
+    }
+
+    emit Locked(msg.sender, lockOrBurnIn.amount);
+
+    return Pool.LockOrBurnOutV1({
+      destTokenAddress: getRemoteToken(lockOrBurnIn.remoteChainSelector),
+      destPoolData: _encodeLocalDecimals()
+    });
+  }
+
+  /// @notice Release tokens from the pool to the recipient
+  /// @dev The _validateReleaseOrMint check is an essential security check
+  function releaseOrMint(
+    Pool.ReleaseOrMintInV1 calldata releaseOrMintIn
+  ) external virtual override returns (Pool.ReleaseOrMintOutV1 memory) {
+    _validateReleaseOrMint(releaseOrMintIn);
+
+    // Calculate the local amount
+    uint256 localAmount =
+      _calculateLocalAmount(releaseOrMintIn.amount, _parseRemoteDecimals(releaseOrMintIn.sourcePoolData));
+
+    // If the message comes from a chain with siloed funds, update internal accounting before releasing tokens.
+    if (s_siloedChainSelectors[releaseOrMintIn.remoteChainSelector]) {
+      s_lockedTokensByChainSelector[releaseOrMintIn.remoteChainSelector] -= localAmount;
```

Comment:
probably good to mention this underflows if there isn't enough to release and with that it protects the other lanes? 

On that note: how does this protect the siloed funds from being given to a non-soloed lane?